### ISSUE TITLE
Easy to hang it (and can't CTRL+C hangs)

### ISSUE THREAD

**Issue by simonmichael** [CrystalSplitter/ghcitui#49]
## Description

I say easy, but I don't have a very precise recipe. 

I set a bunch of breakpoints on successive lines, I stepped into a function, I switched to the REPL window, I tried to evaluate something which maybe hit another breakpoint (?) The output looked like this:
```
ghci> clicmdarg
│Stopped in Hledger.Cli.main.(...), hledger/Hledger/Cli.hs:243:59-87
│_result :: (String, [String], [String]) = _
│cliargs :: [String] = []
```
Then I tried to evaluate the same value again, and the TUI hung.
```
ghci> clicmdarg
```

### Expected behaviour

Not hanging.

### Environment

On mac, using stack, ghc 9.10.1, in the hledger project.

**Comment by simonmichael** [comment]
PS in this state it's also unresponsive to CTRL-c, I had to kill it.

**Comment by CrystalSplitter** [comment]
Thanks for the bug! I see you probably mean your https://github.com/simonmichael/hledger? Is this always reproducible? I'm afraid I'm not really familiar with stack projects, as everything I have worked with so far has been in `cabal`, so you'll have to help me run these targets.

As for the bug itself... This is interesting. Are you sure this is not a GHCi issue? I encountered a lot of problems with GHCi doing this in my evaluation. I would like to be more involved with fixing GHCi personally, but it's currently outside of my available time and energy resources.

**Comment by simonmichael** [comment]
Here's a repro:

- git clone the ghcitui repo
- `ghcitui --cmd 'cabal repl ghcitui:exe:ghcitui'`
- select Main.hs, and set a breakpoint
- start the program with :main, hit the breakpoint
- in the repl, type :main again -> hangs. I'm guessing this similar to the other hang I saw.

**Comment by CrystalSplitter** [comment]
I see. So this reoccurs with GHCi as well. And this particular reproducer "makes sense" as it creates an infinite loop. You can try it out yourself with a `Hello World!`, without any `ghcitui` involved:

```shell
┌ ~
╰ echo 'main = putStrLn "Hello World!" >> putStrLn "Another line!"' > /tmp/Main.hs
┌ ~
╰ ghci /tmp/Main.hs
GHCi, version 9.4.8: https://www.haskell.org/ghc/  :? for help
[1 of 2] Compiling Main             ( /tmp/Main.hs, interpreted )
Ok, one module loaded.
ghci> :b main
Breakpoint 0 activated at /tmp/Main.hs:1:8-58
ghci> main
Stopped in Main.main, /tmp/Main.hs:1:8-58
_result :: IO () = _
[/tmp/Main.hs:1:8-58] ghci> main
<INFINITE LOOP>
```

This creates an intentional infinite loop, and I'm guessing internally in GHCi it's very similar to:

```hs
ghci> x = x
ghci> x
<INFINITE LOOP>
```

I actually doubt there's a solution to this one in GHCi, at least with its current execution model. Welcome to file a bug with them though at https://gitlab.haskell.org/ghc/ghc/-/issues to see if they have any opinions on this.

**Comment by CrystalSplitter** [comment]
> PS in this state it's also unresponsive to CTRL-c, I had to kill it.

I'm a bit more interested in this case though. There are some GHCi situations this is true, but in theory it's possible I'm not forwarding the interrupt in GHCiTUI. I'm not going to close this bug until I'm sure that's what's wrong.

**Comment by CrystalSplitter** [comment]
Alright, found the CTRL+C problem. It's currently unsupported--I just never added it :).

There's currently no way to call [`Ghcid.interrupt`](https://github.com/ndmitchell/ghcid/blob/master/src/Language/Haskell/Ghcid.hs#L220) in GHCiTUI. This will require moving the Ghci process to its own thread and interact with the handle asynchronously. Seems doable, but beyond what I can fix right this moment.

**Comment by CrystalSplitter** [comment]
Some thoughts on implementation here, as this is going to be a fairly large refactor from "two threaded" to an "N threaded" paradigm.

Fundamentally, we can't update the `AppState` without knowing the current state of execution. But we also can't handle interrupts if we're waiting for the Ghcid subprocess to get back to us. 

We're likely going to have to set up a Brick Channel with a custom `AppEvent` which triggers when the Ghcid subprocess completes. While the subprocess is computing something, we cannot update the TUI itself in any way that would conflict with the resolution of that subprocess. For example, the things we can't do while we wait for resolution:

1. You can't run another command while waiting for the execution of the prior command.
2. You aren't allowed to change the module in the source viewer.
3. You aren't allowed to set breakpoints, enable tracing, continue, step, or any other source viewer operation.

Given that, perhaps the only actions you should be allowed to take are window switching ones, dialogs, and interrupts. Maybe even just interrupts.

Some idle thoughts then...

* You aren't allowed to call `runDaemon*`, you must schedule ops on a thread, and set a callback which sends off an `AppEvent` on a BChan.
* DaemonIO ops can't be `run` directly, you can only schedule operations on them and any waiting is handled by `MVar`'s or `TVar`'s.
* `InterpState` or `AppState` needs some record of the type of command that's currently running, so that we can `case` on the resolution and update the TUI accordingly. This seems ugly and I need to think of a way to handle this.

**Comment by CrystalSplitter** [comment]
I have a working prototype of the above, but it will require a lot of testing to ensure we don't hit any deadlocks. Any release is unlikely this weekend.

**Comment by CrystalSplitter** [comment]
Fixed in https://github.com/CrystalSplitter/ghcitui/pull/54. Further testing is still required for release. I encourage you to make sure it resolves your hanging issues.

**Comment by simonmichael** [comment]
Thanks!

I'm not asking for more work, and I don't have clear repros, but here are some testing notes in case they're helpful.

I set a breakpoint, ran :main, hit the breakpoint, ran :main again.

- The app seems hung, as we'd expect, but now ctl-c interrupts it and returns control to ghci - good.
- In the hung state, the ghci pane still echos input, so it's not clear whether it's busy or not. It would be good to disable echo while it's executing maybe.

On my second attempt at this test, I set breakpoint somewhere else. This time running :main repeatedly doesn't hang, it just says "Stopped in (the breakpoint)" and immediately returns control to the user. I don't know why I saw these two behaviours.

More generally: other than ctl-c interrupting the running program, ctl-c and ctl-z seem to be ignored. Perhaps ideally they would do the usual thing (kill or suspect ghcitui).

**Comment by CrystalSplitter** [comment]
> In the hung state, the ghci pane still echos input, so it's not clear whether it's busy or not. It would be good to disable echo while it's executing, maybe.

Tricky. The only way I could imagine implementing this is that each keypress requires `tryTakeMvar`ing the `_ghciLock`. It would likely make it look unresponsive. I could do some testing for this but if it _appears_ to slow down input, I'm not going to implement it. I notice that the newline isn't being printed before exec, so I should do that...

> On my second attempt at this test, I set breakpoint somewhere else. This time running :main repeatedly doesn't hang, it just says "Stopped in (the breakpoint)" and immediately returns control to the user. I don't know why I saw these two behaviours.

I think this is just how GHCi handles breakpoints. This is probably expected behaviour. The hanging less so :).

> More generally: other than ctl-c interrupting the running program, ctl-c and ctl-z seem to be ignored. Perhaps ideally they would do the usual thing (kill or suspend ghcitui).

Hmm. I don't think I ever plan to support CTRL+C being an actual program interrupt, in the same way that other debuggers do not use this to kill them (e.g. `gdb`, `lldb`, `vim`). I think that would violate the rule of least surprise for the UX. I could print out an error message or do `confirmQuit` though.

CTRL+Z would be... tough to support. It is on the table in terms of "this is desirable", so I'm more likely to implement it than CTRL+C actually killing ghcitui. In theory I could intercept the `SIGTSTP` before it gets to the VTY/Brick Signal Handler, but it's very non-trivial. I'll have to think about what this would look like.

**Comment by CrystalSplitter** [comment]
> In the hung state, the ghci pane still echos input, so it's not clear whether it's busy or not. It would be good to disable echo while it's executing, maybe.

This should be fixed in #58.
